stricter parser for ipv4_from_asc

Sashan opened the pull request:
reject invalid IPv4 addresses in ipv4_from_asc

The old scanf-based parser accepted all kinds of invalid inputs like: "1.2.3.4.5"
"1.2.3.4 "
"1.2.3. 4"
" 1.2.3.4"
"1.2.3.4."
"1.2.3.+4"
"1.2.3.4.example.test"
"1.2.3.01"
"1.2.3.0x1"
Thanks to Amir Mohamadi for pointing this out.

<!--
Thank you for your pull request. Please review these requirements:

Contributors guide: https://github.com/openssl/openssl/blob/master/CONTRIBUTING.md

Other than that, provide a description above this comment if there isn't one already

If this fixes a GitHub issue, make sure to have a line saying 'Fixes #XXXX' (without quotes) in the commit message.
-->

##### Checklist
<!-- Remove items that do not apply. For completed items, change [ ] to [x]. -->
- [ ] documentation is added or updated
- [x] tests are added or updated

t8m commented on crypto/x509/v3_utl.c:
```diff
@@ -1151,23 +1151,58 @@ int ossl_a2i_ipadd(unsigned char *ipout, const char *ipasc)
-{
-    const char *p;
-    int a0, a1, a2, a3, n;
+/* get_ipv4_component consumes one IPv4 component, terminated by either '.' or
+ * the end of the string, from |*str|. On success, it returns one, sets |*out|
```
Please reformat the comments. Also there are C++ style comments in the tests.

t8m commented on crypto/x509/v3_utl.c:
```diff
@@ -1151,23 +1151,58 @@ int ossl_a2i_ipadd(unsigned char *ipout, const char *ipasc)
+
+static int ipv4_from_asc(unsigned char *v4, const char *in)
+{
+    if (!get_ipv4_component(&v4[0], &in) || !get_ipv4_dot(&in) ||
+        !get_ipv4_component(&v4[1], &in) || !get_ipv4_dot(&in) ||
+        !get_ipv4_component(&v4[2], &in) || !get_ipv4_dot(&in) ||
+        !get_ipv4_component(&v4[3], &in) || *in != '\0') {
```
This needs also code style reformat

t8m commented on test/x509_test.c:
```diff
@@ -114,6 +117,122 @@ static int test_x509_crl_tbs_cache(void)
+    while (test_samples[i].inp[0]) {
+        memset(buf, 0, BUF_LEN);
+        rc = ossl_a2i_ipadd(buf, test_samples[i].inp);
+        if (((rc == 0) && (test_samples[i].out[0] == '\0')) ||
+            ((rc != 0) && (memcmp(buf, test_samples[i].out, BUF_LEN) == 0))) {
```
and here `||` should be on the beginning of a line

Sashan commented:
new version of PR
   - addresses all nits pointed out by @t8m.
   - the test has been moved from `test/x509_test.c` to `test/x509_internal_test.c`, this solved linker failures on many build targets on github. the earlier version of my test used to call `ossl_a2i_ipadd()` directly, the new version calls `ossl_a2i_ipadd()` via `a2i_IPADDRESS()` function.
   - also note existing tests in `test/x509_internal_test.c` had to be adjusted. earlier code allowed IP addresses with leading spaces `  1.2.3.4` (or `1.2.3.4  ` too). This is an invalid input now.

kroeckx commented on test/x509_internal_test.c:
```diff
@@ -58,22 +58,92 @@ static IP_TESTDATA a2i_ipaddress_tests[] = {
+    {"1.2", NULL, 0 },
+    {"1.2.", NULL, 0 },
+    {"1.2.3", NULL, 0 },
+    {"1.2.3.", NULL, 0 },
```
Some of those are valid IPv4 addresses:
- "1" is "0.0.0.1"
- "1.2" is "1.0.0.2" (and so 127.1 is 127.0.0.1)
- "1.2.3" is "1.2.0.3"

I'm not saying we should accept them, but they are valid and supported by almost all tools.

Sashan replied:
This is a valid point. The current implementation in openssl 3.3 considers those IP addresses as invalid. I'm not sure if we should change the behavior in this PR. I think this needs some more thought. I prefer to keep behavior same. The code in question is part of x509 subsystem. The reason I want to keep 1.2 input refused is it may be source of confusion. Consider situation as follows:
   - the address entered as 1.2 in ASCIIZ form
   -  it gets converted to 1.0.0.2 binary and encoded using ber/der
   - converting the binary value back to ASCIIZ will give 1.0.0.2 which is different to what user entered.

davidben replied:
+1 to what @Sashan said. If OpenSSL wants to switch to accepting those syntaxes, I think that's best done separately. I wrote those tests because test coverage was generally lacking.

That said, what "is" a valid IPv4 address is very context-dependent. If you ask the IETF URL RFC, the answer is [no](https://www.rfc-editor.org/rfc/rfc3986.html#page-20).

But if you ask the WHATWG URL standard, those are allowed, but then you're also allowed to write components in [hex or octal](https://url.spec.whatwg.org/#concept-ipv4-parser), which OpenSSL has also never supported.

If you ask [RFC 952](https://www.rfc-editor.org/rfc/rfc952), cited by [RFC 1123](https://www.rfc-editor.org/rfc/rfc1123#page-13), you have to have four decimal numbers, so the answer is again no.

Given OpenSSL's previous behavior, the more narrow syntax seems to be the best path here.

openssl-machine commented:
This pull request is ready to merge

t8m commented:
Merged to the master branch. Thank you.
